### PR TITLE
Life cycle additions

### PR DESCRIPTION
**Pull request by Kantanomo**
Created c_game_life_cycle_manager and other classes that the game uses to control the life_cycle.

Example code in KantTesting to do direct connection to lobbies, that will be used in the future for Discord invites.

### CODE REVIEW

File: xlive/Blam/Engine/Networking/Session/NetworkSession.cpp [hunk 1]
```diff
@@ -258,4 +258,25 @@ bool network_session_interface_set_local_user_character_type(int user_index, e_c
...
+	{
+		*out_identifier = user_properties->network_user_identifier;
+		return true;
+	}
+	return false;
+}
+
+bool __cdecl network_session_interface_get_local_user_properties_out(int32 user_index, int32* out_controller_index, s_player_properties* out_properties, int32* out_player_voice, int32* out_player_text_chat)
+{
+	return INVOKE(0x1B10E0, 0, network_session_interface_get_local_user_properties_out, user_index, out_controller_index, out_properties, out_player_voice, out_player_text_chat);
```
Comment (Berthalamew) [inline xlive/Blam/Engine/Networking/Session/NetworkSession.cpp]: add server offset

File: xlive/Blam/Engine/Networking/Session/NetworkSession.cpp [hunk 2]
```diff
@@ -258,4 +258,25 @@ bool network_session_interface_set_local_user_character_type(int user_index, e_c
...
+}
+
+void __cdecl network_session_init_session(int unk, char unk_2)
+{
+	INVOKE(0x1B54CF, 0, network_session_init_session, unk, unk_2);
```
Comment (Berthalamew) [inline xlive/Blam/Engine/Networking/Session/NetworkSession.cpp]: add server offset

File: xlive/Blam/Engine/Networking/logic/life_cycle_manager.cpp [hunk 3]
```diff
@@ -1,10 +1,72 @@
...
+	this->life_cycle_manager = life_cycle_manager;
+	this->life_cycle = life_cycle;
+	this->unk_bool = unk_bool;
+	static_cast<c_game_life_cycle_manager*>(this->life_cycle_manager)->life_cycle_handlers[this->life_cycle] = this;
+}
+
+void c_game_life_cycle_handler_joining::check_joining_capability()
+{
+	typedef void(__cdecl check_joining_capability_t)();
+	check_joining_capability_t* p_check_joining_capability = Memory::GetAddress<check_joining_capability_t*>(0x1AD643);
```
Comment (Berthalamew) [inline xlive/Blam/Engine/Networking/logic/life_cycle_manager.cpp]: Replace with INVOKE and add server offset

File: xlive/Blam/Engine/game/game.cpp [hunk 4]
```diff
@@ -41,6 +43,11 @@ bool game_is_ui_shell(void)
 	return game_options_get()->game_mode == _game_mode_ui_shell;
 }
 
+void __cdecl game_shell_set_in_progress()
+{
+	INVOKE(0x242E5B, 0, game_shell_set_in_progress);
```
Comment (Berthalamew) [inline xlive/Blam/Engine/game/game.cpp]: add server offset

File: xlive/Blam/Engine/game/game.cpp [hunk 5]
```diff
@@ -68,3 +75,56 @@ void game_options_setup_default_players(int player_count, s_game_options* game_o
 	p_game_options_setup_default_players(player_count, game_options);
 	return;
 }
+
+void __cdecl reset_global_player_counts()
+{
+	INVOKE(0xB8B9, 0, reset_global_player_counts);
```
Comment (Berthalamew) [inline xlive/Blam/Engine/game/game.cpp]: add server offset

File: xlive/H2MOD/Utils/Utils.cpp [hunk 6]
```diff
@@ -730,3 +730,91 @@ std::string ByteToHexStr(const uint8_t* buffer, size_t size) {
 	}
 	return str.str();
 }
+
+std::string base64_encode(const std::vector<unsigned char>& binaryData) {
```
Comment (Berthalamew) [inline xlive/H2MOD/Utils/Utils.cpp]: These functions are duplicates of the ones in Base64.cpp

**Review by Berthalamew — CHANGES_REQUESTED** [review]

File: xlive/H2MOD/Utils/Utils.cpp (hunk 6 above)
Comment (Kantanomo) [reply xlive/H2MOD/Utils/Utils.cpp]: Functions are different, but I moved them into the same files and renamed them to base64_encode_binary and base64_decode_binary.

**Review by Berthalamew — APPROVED** [review]

**Pull request merged**